### ISSUE TITLE
[Bug]: Assertion hit in row-by-row fetcher

### ISSUE THREAD

**Issue by erimatnor** [timescale/timescaledb#4177]
### What type of bug is this?

Crash

### What subsystems and features are affected?

Distributed hypertable

### What happened?

A crash occurred when running Sqlsmith. It is hard to reproduce since Sqlsmith CI job runs use random seeds and it is unclear which seed caused the crash. 

Link to failed CI job: https://github.com/timescale/timescaledb/runs/5568184043?check_suite_focus=true

![sqlsmith-row-fetcher-crash](https://user-images.githubusercontent.com/819732/158771706-190b180c-d43e-468d-97e6-2f1f39d72473.png)

### TimescaleDB version affected

2.7.0-dev (ce3e04a9ec945f8734cc5cdd0a258e8130d5c210)

### PostgreSQL version used

14.2

### What operating system did you use?

Ubuntu 20.04 x86

### What installation method did you use?

Source

### What platform did you run on?

Other

### Relevant log output and stack trace

```bash
Stack trace provided above. (Could not upload png here.)

Sqlsmith log output:

SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:20:42 Seed 2602653
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:25:33 Seed 2569024
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:30:14 Seed 4995376
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:35:22 Seed 9489005
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:40:06 Seed 1690467
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:44:55 Seed 4659337
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:49:09 Seed 9343914
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:53:57 Seed 11901484
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 10:58:25 Seed 7592973
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-16 11:02:24 Seed 4115891

```

### How can we reproduce the bug?

One could possibly run Sqlsmith with seed `4115891`. However, it wasn't possible to reproduce the crash on a local Ubuntu 21.10 installation using this seed.

**Comment by erimatnor** [comment]
Another Sqlsmith run produced the same bug:

```
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:01:25 Seed 15777643
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:05:22 Seed 5000994
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:08:49 Seed 8895015
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:12:02 Seed 15901630
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:15:30 Seed 10257587
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:19:14 Seed 13853837
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:23:58 Seed 10615273
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:27:23 Seed 361714
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:31:16 Seed 376752
SQLsmith 2014d21
Loading types...done.
Loading tables...done.
Loading columns and constraints...done.
Loading operators...done.
Loading routines...done.
Loading routine parameters...done.
Loading aggregates...done.
Loading aggregate parameters...done.
Generating indexes...done.
2022-03-17 20:34:58 Seed 2724980
```


Stack trace:

```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `postgres: runner smith [local] SELECT                                         '.
Program terminated with signal SIGABRT, Aborted.
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
(gdb) #0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
        set = {__val = {4194304, 0, 0, 0, 0, 94199284171552, 
            4611686018427387904, 0, 0, 140114675975386, 0, 94199284171598, 
            128, 0, 140725350505248, 140114670048576}}
        pid = <optimized out>
        tid = <optimized out>
        ret = <optimized out>
#1  0x00007f6efd0f0859 in __GI_abort () at abort.c:79
        save_stage = 1
        act = {__sigaction_handler = {sa_handler = 0x7f6efd283bf7, 
            sa_sigaction = 0x7f6efd283bf7}, sa_mask = {__val = {1, 
              140114670407710, 3, 140725350505428, 12, 140114670407714, 2, 
              7161063386381837328, 7292509979588976951, 140725350505536, 
              7291948136861589116, 94199297811296, 4290875770853815296, 
              140725350505520, 307, 18}}, sa_flags = 747052096, 
          sa_restorer = 0x7f6ef334de7c}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x000055ac7b26afb8 in ExceptionalCondition (
    conditionName=conditionName@entry=0x7f6ef334de7c "fetcher->state.open", 
    errorType=errorType@entry=0x7f6ef333956e "FailedAssertion", 
    fileName=fileName@entry=0x7f6ef334dd40 "/home/runner/work/timescaledb/timescaledb/tsl/src/remote/row_by_row_fetcher.c", lineNumber=lineNumber@entry=307)
    at assert.c:69
No locals.
#3  0x00007f6ef332fdf9 in row_by_row_fetcher_close (df=0x55ac814049a0)
    at /home/runner/work/timescaledb/timescaledb/tsl/src/remote/row_by_row_fetcher.c:307
        fetcher = <optimized out>
#4  row_by_row_fetcher_close (df=0x55ac814049a0)
    at /home/runner/work/timescaledb/timescaledb/tsl/src/remote/row_by_row_fetcher.c:303
        fetcher = 0x55ac814049a0
#5  0x00007f6ef332a828 in data_fetcher_free (df=0x55ac814049a0)
    at /home/runner/work/timescaledb/timescaledb/tsl/src/remote/data_fetcher.c:115
No locals.
#6  0x00007f6ef33110c3 in fdw_scan_end (fsstate=0x55ac7e3443c0)
    at /home/runner/work/timescaledb/timescaledb/tsl/src/fdw/scan_exec.c:340
No locals.
#7  0x000055ac7afd03aa in ExecEndCustomScan (node=0x55ac7e3442a8)
    at nodeCustom.c:122
No locals.
#8  0x000055ac7afbc501 in ExecEndNode (node=<optimized out>)
    at execProcnode.c:690
        __func__ = "ExecEndNode"
#9  0x000055ac7afcdb84 in ExecEndAppend (node=<optimized out>)
    at nodeAppend.c:411
        appendplans = <optimized out>
        nplans = <optimized out>
        i = <optimized out>
#10 0x000055ac7afd03aa in ExecEndCustomScan (node=0x55ac7e3438d0)
    at nodeCustom.c:122
No locals.
#11 0x000055ac7afbc501 in ExecEndNode (node=<optimized out>)
    at execProcnode.c:690
        __func__ = "ExecEndNode"
#12 0x000055ac7afe7d81 in ExecEndNestLoop (node=0x55ac7e3437b8)
    at nodeNestloop.c:380
No locals.
#13 0x000055ac7afbc4f1 in ExecEndNode (node=<optimized out>)
    at execProcnode.c:697
        __func__ = "ExecEndNode"
#14 0x000055ac7afb62f4 in ExecEndPlan (estate=0x55ac810ace90, 
    planstate=<optimized out>) at execMain.c:1416
        subplanstate = <optimized out>
        l__state = {l = <optimized out>, i = 2}
        l = <optimized out>
        l = <optimized out>
        l__state = {l = <optimized out>, i = <optimized out>}
        subplanstate = <optimized out>
#15 standard_ExecutorEnd (queryDesc=0x55ac7cca2fa0) at execMain.c:493
        estate = 0x55ac810ace90
        oldcontext = 0x55ac7baf2360

#16 0x000055ac7af6d574 in PortalCleanup (portal=<optimized out>)
warning: Can't open file /dev/shm/PostgreSQL.3440596380 during file-backed mapping note processing
    at portalcmds.c:299

        saveResourceOwner = 0x55ac7ba8c4b8
warning: Can't open file /dev/shm/PostgreSQL.79589546 during file-backed mapping note processing
        queryDesc = 0x55ac7cca2fa0

#17 0x000055ac7b299238 in PortalDrop (portal=0x55ac7bac77d0, 
warning: Can't open file /dev/shm/PostgreSQL.4202783102 during file-backed mapping note processing
    isTopCommit=<optimized out>) at portalmem.c:502

        __func__ = "PortalDrop"
warning: Can't open file /dev/zero (deleted) during file-backed mapping note processing
#18 0x000055ac7b14a868 in exec_simple_query (

    query_string=0x55ac7ba51da0 "select  \n  sample_0.time as c0, \n  sample_0.time as c1, \n  sample_0.time as c2, \n  sample_0.sensor_id as c3, \n  cast(nullif((select device from public.dist_chunk_copy limit 1 offset 3)\n      ,\n    sam"...) at postgres.c:1224
warning: Can't open file /SYSV001bc83b (deleted) during file-backed mapping note processing
        snapshot_set = <optimized out>
50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
        per_parsetree_context = 0x0
        plantree_list = <optimized out>
        parsetree = <optimized out>
        commandTag = <optimized out>
        qc = {commandTag = CMDTAG_SELECT, nprocessed = 0}
        querytree_list = <optimized out>
        portal = 0x55ac7bac77d0
        receiver = <optimized out>
        format = 0
        parsetree_item__state = {l = 0x55ac7e5164b8, i = 0}
        dest = DestRemote
        oldcontext = <optimized out>
        parsetree_list = 0x55ac7e5164b8
        parsetree_item = <optimized out>
        save_log_statement_stats = false
        was_logged = false
        use_implicit_block = false
        msec_str = "\000\000\000\000\000\000\000\000\000t\022\221\307@\214;\000\000\000\000\000\000\000\000{\353'{\254U\000"
        __func__ = "exec_simple_query"
#19 0x000055ac7b14c63d in PostgresMain (argc=argc@entry=1, 
    argv=argv@entry=0x7ffd2c8725a0, dbname=<optimized out>, 
    username=<optimized out>) at postgres.c:4486
        query_string = 0x55ac7ba51da0 "select  \n  sample_0.time as c0, \n  sample_0.time as c1, \n  sample_0.time as c2, \n  sample_0.sensor_id as c3, \n  cast(nullif((select device from public.dist_chunk_copy limit 1 offset 3)\n      ,\n    sam"...
        firstchar = <optimized out>
        input_message = {
          data = 0x55ac7ba51da0 "select  \n  sample_0.time as c0, \n  sample_0.time as c1, \n  sample_0.time as c2, \n  sample_0.sensor_id as c3, \n  cast(nullif((select device from public.dist_chunk_copy limit 1 offset 3)\n      ,\n    sam"..., len = 2646, maxlen = 4096, cursor = 2646}
        local_sigjmp_buf = {{__jmpbuf = {0, 6224316091460422755, 1748856046, 
              94199297344912, 3, 94199292110344, 6224316091510754403, 
              199171750102764643}, __mask_was_saved = 1, __saved_mask = {
              __val = {4194304, 140725350518737, 0, 0, 140725350507904, 
                140114668598976, 4290875770853815296, 206158430232, 
                94199289213267, 206158430240, 140725350507824, 
                140725350507632, 4290875770853815296, 94199297307952, 0, 6}}}}
        send_ready_for_query = false
        idle_in_transaction_timeout_enabled = false
        idle_session_timeout_enabled = false
        __func__ = "PostgresMain"
#20 0x000055ac7b0b9df3 in BackendRun (port=<optimized out>, 
    port=<optimized out>) at postmaster.c:4530
        av = {0x55ac7b2d07b2 "postgres", 0x0}
        ac = 1
        av = {<optimized out>, <optimized out>}
        ac = <optimized out>
#21 BackendStartup (port=<optimized out>) at postmaster.c:4252
        bn = <optimized out>
        pid = <optimized out>
        bn = <optimized out>
        pid = <optimized out>
        __func__ = "BackendStartup"
        __errno_location = <optimized out>
        __errno_location = <optimized out>
        save_errno = <optimized out>
        __errno_location = <optimized out>
        __errno_location = <optimized out>
#22 ServerLoop () at postmaster.c:1745
        port = <optimized out>
        i = <optimized out>
        rmask = {fds_bits = {128, 0 <repeats 15 times>}}
        selres = <optimized out>
        now = <optimized out>
        readmask = {fds_bits = {224, 0 <repeats 15 times>}}
        nSockets = 8
        last_lockfile_recheck_time = 1647547875
        last_touch_time = 1647547273
        __func__ = "ServerLoop"
#23 0x000055ac7b0bad79 in PostmasterMain (argc=argc@entry=6, 
    argv=argv@entry=0x55ac7ba49bb0) at postmaster.c:1417
        opt = <optimized out>
        status = <optimized out>
        userDoption = <optimized out>
        listen_addr_saved = true
        i = <optimized out>
        output_config_variable = <optimized out>
        __func__ = "PostmasterMain"
#24 0x000055ac7adf02a2 in main (argc=6, argv=0x55ac7ba49bb0) at main.c:209
        do_check_root = <optimized out>
(gdb) quit
           PID: 13040 (postgres)
           UID: 1001 (runner)
           GID: 121 (docker)
        Signal: 6 (ABRT)
     Timestamp: Thu 2022-03-17 20:14:46 UTC (23min ago)
  Command Line: postgres: runner smith [local] SELECT
    Executable: /home/runner/postgresql/bin/postgres
 Control Group: /system.slice/runner-provisioner.service
          Unit: runner-provisioner.service
         Slice: system.slice
       Boot ID: 5eb753978ba1447db30252ef112948f7
    Machine ID: 6fe698402f044602bf038833560abb94
      Hostname: fv-az397-458
       Storage: /var/lib/systemd/coredump/core.postgres.1001.5eb753978ba1447db30252ef112948f7.13040.1647548086000000000000.lz4
       Message: Process 13040 (postgres) of user 1001 dumped core.
                
                Stack trace of thread 13040:
                #0  0x00007f6efd11103b __GI_raise (libc.so.6 + 0x4303b)
                #1  0x00007f6efd0f0859 __GI_abort (libc.so.6 + 0x22859)
                #2  0x000055ac7b26afb8 n/a (/home/runner/postgresql/bin/postgres + 0x5d3fb8)
```

**Comment by svenklemm** [comment]
Fixed by #4516
